src/NHibernate.Validator/Engine/ClassValidator.cs

```diff
@@ -230,9 +236,23 @@ private void InitValidator(System.Type clazz, IDictionary<System.Type, IClassVal
 							});
 					}
 				}
+
+				if (membersToValidateFor1Memeber.Count > 0)
+				{
+					membersToValidate.AddRange(membersToValidateFor1Memeber);
+					membersToValidateFor1Memeber.Clear();
+				}
 			}
 		}
 
+		private void RemoveSameTypeValidatorWithIntersectingTags(List<Member> membersToValidateFor1Memeber, ITagableRule newAttr)
+		{
+				membersToValidateFor1Memeber.RemoveAll(x => x.ValidatorDef.Validator.GetType() == newAttr.GetType()
+				                                            && ((x.ValidatorDef.Tags == null && newAttr.TagCollection == null)
+				                                                || x.ValidatorDef.Tags.Intersect(newAttr.TagCollection).Any()
+				                                            ));

```

Comment:
This need renaming, rewriting and reformatting, for maintainability and correctness. Please use semantic naming (what is it meant to do, rather than what it technically does).

By example:
```c#
private void RemoveOverridenValidators(System.Type clazz, List<Member> memberValidators, ITagableRule newValidator)
{
	var newValidatorType = newValidator.GetType();
	var newValidatorIsUntagged = newValidator.TagCollection.Count == 0;
	var removedCount = memberValidators
		.RemoveAll(
			v => v.ValidatorDef.Validator.GetType() == newValidatorType && (
				(v.ValidatorDef.Tags == null || v.ValidatorDef.Tags.Count == 0) && newValidatorIsUntagged ||
				v.ValidatorDef.Tags != null && !newValidatorIsUntagged && v.ValidatorDef.Tags.Intersect(newValidator.TagCollection).Any()));
	if (removedCount > 0)
	{
#if NETFX
	log.Debug(
		"Validators of the same type having some tags in common (or both without tags) found on a " +
			"member, keeping only the last defined: class " + clazz.FullName + ", member " + v.Getter.Name +
			", last validator " + newValidator);
#else
	Log.Debug(
		"Validators of the same type having some tags in common (or both without tags) found on a " +
			"member, keeping only the last defined: class {0}, member: {1}, last validator {2}",
		clazz.FullName, v.Getter.Name, newValidator);
#endif
	}
}
```
I have removed the null check on `TagCollection` since it is never checked everywhere else and that current implementations never yields null. I also have added a log for helping diagnosing rule overrides.

But shouldn't it check `AllowsMultipleWithIntersectingTags` before removing?